### ISSUE TITLE
A question about loss computation in Multi GPU setting?

### ISSUE THREAD

**Issue by Remember2018** [tensorpack/tensorpack#1350]
Hi Yuxin, I'm trying to port the FCOS (https://github.com/tianzhi0549/FCOS) based on Tensorpack framework. I have a question on the loss computation in multi-gpu setting. 

In FCOS, the loss computation is like this:

https://github.com/tianzhi0549/FCOS/blob/b516eb54dc069d9c0fadb0d51eeead7dd1bb4910/fcos_core/modeling/rpn/fcos/loss.py#L250

        num_gpus = get_num_gpus()
        # sync num_pos from all gpus
        total_num_pos = reduce_sum(pos_inds.new_tensor([pos_inds.numel()])).item()
        num_pos_avg_per_gpu = max(total_num_pos / float(num_gpus), 1.0)

        cls_loss = self.cls_loss_func(
            box_cls_flatten,
            labels_flatten.int()
        ) / num_pos_avg_per_gpu

It seems the `reduce_sum` is computed in advance, then the loss is averaged by the sum of the values from all GPUs. To my understanding, in the framework of Tensorpack, the Multi-GPU is automatically set and we just need to write the `build_graph` function to construct the graph. 
My question is, in terms of FCOS in Tensorpack, do we need to use the `reduce_sum` in the loss computation in `build_graph` and then manually compute the loss? Did the Tensorpack do this work? Shall we just use the `num_pos` from one GPU, and Tensorpack will do this loss average automatically? I'm not sure if I understood correctly, could you please give some advice? Thanks.

**Comment by ppwwyyxx** [comment]
`build_graph` returns a scalar loss, so how to obtain it is up to you.
The loss from each `build_graph` is either averaged or added based on the `average` option in https://tensorpack.readthedocs.io/modules/train.html#tensorpack.train.SyncMultiGPUTrainerReplicated.

**Comment by Remember2018** [comment]
@ppwwyyxx Yeah, got it. Thanks a lot for your kind comment!
